linera-execution/src/runtime.rs

```diff
@@ -959,21 +963,27 @@ impl<UserInstance> BaseRuntime for SyncRuntimeInternal<UserInstance> {
             cfg!(feature = "unstable-oracles"),
             ExecutionError::UnstableOracle
         );
-        let response =
-            if let Some(response) = self.transaction_tracker.next_replayed_oracle_response()? {
-                match response {
-                    OracleResponse::Service(bytes) => bytes,
-                    _ => return Err(ExecutionError::OracleResponseMismatch),
-                }
-            } else {
-                let context = QueryContext {
-                    chain_id: self.chain_id,
-                    next_block_height: self.height,
-                    local_time: self.local_time,
-                };
-                let sender = self.execution_state_sender.clone();
-                ServiceSyncRuntime::new(sender, context).run_query(application_id, query)?
+        let QueryOutcome {
+            response,
+            operations,
+        } = if let Some(response) = self.transaction_tracker.next_replayed_oracle_response()? {
+            match response {
+                OracleResponse::Service(bytes) => QueryOutcome {
+                    response: bytes,
+                    operations: vec![],
+                },
+                _ => return Err(ExecutionError::OracleResponseMismatch),
+            }
+        } else {
+            let context = QueryContext {
+                chain_id: self.chain_id,
+                next_block_height: self.height,
+                local_time: self.local_time,
             };
+            let sender = self.execution_state_sender.clone();
+            ServiceSyncRuntime::new(sender, context).run_query(application_id, query)?
+        };
+        self.scheduled_operations.extend(operations);
```

Comment:
This is probably only needed in the `else` case above. Or actually, should we check whether we're currently in a service query, and never add to `scheduled_operations` if we're in an operation?

Reply:
I was thinking about changing the `transaction_tracker` in a separate PR, so that it is an `enum` with  the current `TransactionTracker` and a new `QueryTracker` or something, so that it would then schedule operations only in queries (and fail otherwise) and maybe only replay oracles if it's being called from a contract. I need to think more about this, so I left it for later :sweat_smile:

Reply:
Makes sense! :+1: